include/net/http_client.h

```diff
@@ -0,0 +1,289 @@
+
+	/** User supplied buffer where received data is stored */
+	u8_t *recv_buf;
+
+	/** Length of the user supplied receive buffer */
+	size_t recv_buf_len;
+
+	/** The URL for this request, for example: /index.html */
+	const char *url;
+
+	/** The HTTP protocol, for example "HTTP/1.1" */
+	const char *protocol;
+
+	/** The HTTP header fields (application specific)
+	 * The Content-Type may be specified here or in the next field.
+	 * Depending on your application, the Content-Type may vary, however
+	 * some header fields may remain constant through the application's
+	 * life cycle. This is a NULL terminated list of header fields.
+	 */
+	const char **header_fields;
+
+	/** The value of the Content-Type header field, may be NULL */
+	const char *content_type_value;
+
+	/** Hostname to be used in the request */
+	const char *host;
+
+	/** User supplied callback function to call when payload
+	 * needs to be sent. This can be NULL in which case the payload field
+	 * in http_request is used. The idea of this payload callback is to
+	 * allow user to send more data that is practical to store in allocated
+	 * memory.
+	 */
+	http_payload_cb_t payload_cb;
+
+	/** Payload, may be NULL */
+	const char *payload;
+
+	/** Payload length, may be 0. Only used if payload field is not NULL */
+	size_t payload_len;
+
+	/** User supplied callback function to call when optional headers need
+	 * to be sent. This can be NULL, in which case the optional_headers
+	 * field in http_request is used. The idea of this payload callback is
+	 * to allow user to send more data that is practical to store in
+	 * allocated memory.
+	 */
+	http_header_cb_t optional_headers_cb;
+
+	/** A NULL terminated list of any optional headers that
+	 * should be added to the HTTP request. May be NULL.
+	 * If the optional_headers_cb is specified, then this field is ignored.
+	 * Note that there are two similar fields that contain headers,
+	 * the header_fields above and this optional_headers. This is done
+	 * like this to support Websocket use case where Websocket will use
+	 * header_fields variable and any optional application specific
+	 * headers will be placed into this field.
+	 */
+	const char **optional_headers;
+};
+
+/**
+ * @brief Do a HTTP request. The callback is called when data is received
+ * from the HTTP server. The caller must have created a connection to the
+ * server before calling this function so connect() call must have be done
+ * successfully for the socket.
+ *
+ * @param sock Socket id of the connection.
+ * @param req HTTP request information
+ * @param timeout Max timeout to wait for the data. The timeout value cannot be
+ *        0 as there would be no time to receive the data.
+ * @param user_data User specified data that is passed to the callback.
+ *
+ * @return <0 if error, >=0 amount of data sent to the server
+ */
+int http_client_req(int sock, struct http_request *req,
+		    s32_t timeout, void *user_data);
```

Comment:
Is there a reason why `user_data` is passed as a separate param and is not part of `struct http_request`? (Just in case, I'm asking, not calling for change.)

Reply:
Many APIs in networking side etc. have typically timeout and user_data parameters in function calls. Just followed this style here.

Reply:
> Just followed this style here.

Makes sense.